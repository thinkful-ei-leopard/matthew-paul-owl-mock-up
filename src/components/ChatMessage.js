import React from "react";
import STORE from "../STORE";

function renderNonMessage(participant, type, timestamp) {
    let action;
    switch (type) {
        case 'thumbs-up':
            action = ' gave a thumbs up!';
            break;
        case 'thumbs-down':
            action = ' gave a thumbs down!';
            break;
        case 'raise-hand':
            action = ' raised their hand!';
            break;
        case 'clap':
            action = ' clapped!';
            break;
        case 'join':
            action = ' joined!';
            break;
        case 'leave':
            action = ' left!';
            break;
        case 'join-stage':
            action = ' joined the stage!';
            break;
        case 'leave-stage':
            action = ' left the stage!';
            break;
        default:
            action = '';
    }
  return ( 
    <div className="Chat-Action">
       <strong>{participant.name}</strong>{action}
    </div>
  );
}

function ChatMessage(props) {
  let participant = STORE.findParticipantById(props.participantId);
  if (props.type !== "message") {
    return renderNonMessage(participant, props.type, props.timestamp);
  } else {
      let date = new Date(props.timestamp);
    return (
      <div className="Chat-Entry">
        <img className="Participant-Avatar" src={participant.avatar} alt="Avatar" />
        <span className="Participant-Name">{participant.name}</span>
        <span className="Chat-Timestamp">{date.toLocaleTimeString()}</span>
        <p className="Chat-Message">{props.message}</p>
      </div>
    );
  }
}

export default ChatMessage;

/**
 * Props:
 * participantId: Number
 * type: String(message, thumbs up, thumbs down, clap, etc)
 * message: String
 * timestamp: Number of milliseconds sense epoch
 */
