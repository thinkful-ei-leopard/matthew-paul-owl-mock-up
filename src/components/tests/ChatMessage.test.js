import React from 'react';
import ReactDOM from 'react-dom';
import ChatMessage from '../ChatMessage';
import renderer from 'react-test-renderer'; 

/**
 * Tests for regular message
 */
it('renders without crashing', () => {
  // first create a DOM element to render the component into
  const div = document.createElement('div');

  // render the component, this is the actual test, if something is wrong it will fail here
  ReactDOM.render(<ChatMessage key={4}
  participantId={1}
  type='message'
  message='Hello'
  timestamp={1548852544247}/>, div);

  // clean up code
  ReactDOM.unmountComponentAtNode(div);
});

/**
 * Tests for action message
 */
it('renders without crashing', () => {
    // first create a DOM element to render the component into
    const div = document.createElement('div');
  
    // render the component, this is the actual test, if something is wrong it will fail here
    ReactDOM.render(<ChatMessage key={1}
    participantId={1}
    type='thumbs-up'
    timestamp={1548852544247}/>, div);
  
    // clean up code
    ReactDOM.unmountComponentAtNode(div);
  });

/**
 * Tests for regular message
 */
it('renders the UI as expected', () => {
  const tree = renderer
  .create(<ChatMessage key={2}
      participantId={1}
      type='message'
      message='Hello'
      timestamp={1548852544247}/>)
    .toJSON();
  expect(tree).toMatchSnapshot();  
  });

  /**
   * Tests for action message
   */
  it('renders the UI as expected', () => {
    const tree = renderer
    .create(<ChatMessage key={3}
        participantId={1}
        type='thumbs-up'
        timestamp={1548852544247}/>)
      .toJSON();
    expect(tree).toMatchSnapshot();  
    });