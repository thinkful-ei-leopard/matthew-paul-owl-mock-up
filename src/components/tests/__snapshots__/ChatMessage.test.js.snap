// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the UI as expected 1`] = `
<div
  className="Chat-Entry"
>
  <img
    alt="Avatar"
    className="Participant-Avatar"
    src="https://robohash.org/itaquevoluptatumrerum.jpg?size=200x200&set=set1"
  />
  <span
    className="Participant-Name"
  >
    Koren Templeton
  </span>
  <span
    className="Chat-Timestamp"
  >
    1548852544247
  </span>
  <p
    className="Chat-Message"
  >
    Hello
  </p>
</div>
`;

exports[`renders the UI as expected 2`] = `
<div
  className="Chat-Entry"
>
  <span
    className="Chat-Action"
  >
    Koren Templeton
     gave a thumbs up!
  </span>
</div>
`;
